src/baconian.rs

```diff
@@ -0,0 +1,467 @@
+    if distinct {
+        if DISTINCT_CODES.contains_key(key.to_uppercase().as_str()) {
+            code.push_str(DISTINCT_CODES.get(key.to_uppercase().as_str()).unwrap());
+        }
+    } else {
+        // Need to handle 'I'/'J' and 'U'/'V'
+        let mut key_upper = key.to_uppercase();
+
+        match key_upper.as_str() {
+            "J" => key_upper = "I".to_string(),
+            "U" => key_upper = "V".to_string(),
+            _ => {}
+        }
+        if TRAD_CODES.contains_key(key_upper.as_str()) {
+            code.push_str(TRAD_CODES.get(key_upper.as_str()).unwrap());
+        }
+    }
+
+    code
+}
+
+/// Gets the key (the source character) for a given cipher text code
+fn get_key(distinct: bool, code: &str) -> String {
+    let mut key = String::new();
+
+    let codes = if distinct {
+        DISTINCT_CODES.iter()
+    } else {
+        TRAD_CODES.iter()
+    };
+    for (_key, val) in codes {
+        if val == &code {
+            key.push_str(_key);
+        }
+    }
+    key
+}
+
+/// This struct is created by the `new()` method. See its documentation for more.
+pub struct Baconian {
+    distinct: bool,
+    decoy_text: String,
+}
+
+impl Cipher for Baconian {
+    type Key = (bool, Option<String>);
+    type Algorithm = Baconian;
+
+    /// Initialise a Baconian cipher
+    ///
+    /// The `key` tuple maps to the following:
+    ///     `(bool, Option<str>) =
+    ///         (distinct, decoy_text)`.
+    ///
+    /// Where ...
+    ///
+    /// * The encoding will be distinct for all alphabetical characters, or classical
+    ///     where I, J, U and V are mapped to the same value pairs
+    /// * An optional decoy message that will will be used to hide the message -
+    ///     default is boilerplate "Lorem ipsum" text.
+    ///
+    fn new(key: (bool, Option<String>)) -> Result<Baconian, &'static str> {
+        Ok(Baconian {
+            distinct: key.0,
+            decoy_text: key.1.unwrap_or_else(|| String::from(DEFAULT_DECOY)),
+        })
+    }
+
+    /// Encrypt a message using the Baconian cipher
+    ///
+    /// * The message to be encrypted can only be ~18% of the decoy_text as each character
+    ///     of message is encoded by 5 encoding characters `AAAAA`, `AAAAB`, etc.
+    /// * The italicised ciphertext is then hidden in a decoy text, where, for each 'B'
+    ///     in the ciphertext, the character is italicised in the decoy_text.
+    ///
+    /// # Examples
+    /// Basic usage:
+    ///
+    /// ```
+    /// use cipher_crypt::{Cipher, Baconian};
+    ///
+    /// let b = Baconian::new((false, None)).unwrap();
+    /// let message = "Hello";
+    /// let cipher_text = "Lo𝘳𝘦𝘮 ip𝘴um d𝘰l𝘰r s𝘪t 𝘢me𝘵, 𝘯e 𝘵";
+    ///
+    /// assert_eq!(cipher_text, b.encrypt(message).unwrap());
+    /// ```
+    ///
+    fn encrypt(&self, message: &str) -> Result<String, &'static str> {
+        let mut non_alphas = 0; // A counter for non_alphas
+
+        for c in self.decoy_text.chars() {
```

Comment:
You could tricky with iterators here... something like:
```
self.decoy_text.chars().for_each(|c| {if !c.is_alphabetic() {non_alphas += 1}});
```
I don't know what the exact call would look like but it's worth exploring if you haven't done so already.

Reply:
Okay, I'll look into it. Personally, I'm not a great fan of (pythonic) idioms like this as it makes code readability worse imho. I get it for interpreted languages where concise code gives performance gains, like `python` and `Javascript`, but compiled languages, it just makes for hard to maintain code. `</rant>` :smile:

From a learning perspective, though, it's good to use, so when I see it I'll be able to read it (maybe)...

Reply:
Lunchtime look: your snippet is sound and does the same as original. But, actually not gaining anything there - `rustfmt` will reformat over the same number of lines... and don't actually make any savings.

From a learner perspective, I can see what you are thinking. I'm sure I (we) can find a better example to learn that syntax.

Reply:
Sure. I don't feel to strongly either way and I haven't really used streams extensively in this project, so we can keep it as you have for now.